I'm getting various errors, such as:

```
error[E0599]: the method `safe_to_jsval` exists for reference `&str`, but its trait bounds were not satisfied
   --> components/script/dom/xmlhttprequest.rs:935:24
    |
935 |                     "".safe_to_jsval(cx, rval);
    |                        ^^^^^^^^^^^^^
    |
    = note: the following trait bounds were not satisfied:
            `str: Sized`
            which is required by `str: SafeToJSValConvertible`
            `&str: ToJSValConvertible`
            which is required by `&str: SafeToJSValConvertible`
```

```
error[E0599]: the method `safe_to_jsval` exists for array `[bool; 3]`, but its trait bounds were not satisfied
    --> components/script/dom/webglrenderingcontext.rs:4236:68
     |
4236 |                 uniform_get(triple, WebGLCommand::GetUniformBool3).safe_to_jsval(cx, rval);
     |                                                                    ^^^^^^^^^^^^^
     |
     = note: the following trait bounds were not satisfied:
             `[bool; 3]: ToJSValConvertible`
             which is required by `[bool; 3]: SafeToJSValConvertible`
             `[bool]: Sized`
             which is required by `[bool]: SafeToJSValConvertible`
```

Does this mean that I should replace the single `impl<T: ToJSValConvertible> SafeToJSValConvertible for T` with one implementation for each type for which `ToJSValConvertible` is [implemented](https://github.com/servo/mozjs/blob/128e1e230a0155ed9395c718184f34bddc045860/mozjs/src/conversions.rs), including one for `&'_ T`?

@mrobinson

--- a/components/script/dom/bindings/constructor.rs
+++ b/components/script/dom/bindings/constructor.rs
@@ -6,11 +6,11 @@ use std::ptr;
 
 use html5ever::interface::QualName;
 use html5ever::{LocalName, local_name, ns};
-use js::conversions::ToJSValConvertible;
 use js::glue::{UnwrapObjectDynamic, UnwrapObjectStatic};
 use js::jsapi::{CallArgs, CurrentGlobalOrNull, JSAutoRealm, JSObject};
 use js::rust::wrappers::{JS_SetPrototype, JS_WrapObject};
 use js::rust::{HandleObject, MutableHandleObject, MutableHandleValue};
+use script_bindings::conversions::SafeToJSValConvertible;
 use script_bindings::interface::get_desired_proto;
 
 use super::utils::ProtoOrIfaceArray;
@@ -242,7 +242,7 @@ fn html_constructor(
 
         JS_SetPrototype(*cx, element.handle(), prototype.handle());
 
-        result.to_jsval(*cx, MutableHandleValue::from_raw(call_args.rval()));
+        result.safe_to_jsval(cx, MutableHandleValue::from_raw(call_args.rval()));
     }
     Ok(())
 }

--- a/components/script/dom/readablestream.rs
+++ b/components/script/dom/readablestream.rs
@@ -12,7 +12,7 @@ use base::id::{MessagePortId, MessagePortIndex};
 use constellation_traits::MessagePortImpl;
 use dom_struct::dom_struct;
 use ipc_channel::ipc::IpcSharedMemory;
-use js::conversions::ToJSValConvertible;
+use script_bindings::conversions::SafeToJSValConvertible;
 use js::jsapi::{Heap, JSObject};
 use js::jsval::{JSVal, ObjectValue, UndefinedValue};
 use js::rust::{
@@ -1595,7 +1595,6 @@ impl ReadableStream {
     }
 
     /// <https://streams.spec.whatwg.org/#readable-stream-cancel>
-    #[allow(unsafe_code)]
     pub(crate) fn cancel(
         &self,
         cx: SafeJSContext,
@@ -1613,12 +1612,10 @@ impl ReadableStream {
         // If stream.[[state]] is "errored", return a promise rejected with stream.[[storedError]].
         if self.is_errored() {
             let promise = Promise::new(global, can_gc);
-            unsafe {
-                rooted!(in(*cx) let mut rval = UndefinedValue());
-                self.stored_error.to_jsval(*cx, rval.handle_mut());
-                promise.reject_native(&rval.handle(), can_gc);
-                return promise;
-            }
+            rooted!(in(*cx) let mut rval = UndefinedValue());
+            self.stored_error.safe_to_jsval(cx, rval.handle_mut());
+            promise.reject_native(&rval.handle(), can_gc);
+            return promise;
         }
         // Perform ! ReadableStreamClose(stream).
         self.close(can_gc);
@@ -2322,7 +2319,6 @@ impl CrossRealmTransformReadable {
 
     /// <https://streams.spec.whatwg.org/#abstract-opdef-setupcrossrealmtransformwritable>
     /// Add a handler for port’s messageerror event with the following steps:
-    #[allow(unsafe_code)]
     pub(crate) fn handle_error(
         &self,
         cx: SafeJSContext,
@@ -2334,7 +2330,7 @@ impl CrossRealmTransformReadable {
         // Let error be a new "DataCloneError" DOMException.
         let error = DOMException::new(global, DOMErrorName::DataCloneError, can_gc);
         rooted!(in(*cx) let mut rooted_error = UndefinedValue());
-        unsafe { error.to_jsval(*cx, rooted_error.handle_mut()) };
+        error.safe_to_jsval(cx, rooted_error.handle_mut());
 
         // Perform ! CrossRealmTransformSendError(port, error).
         port.cross_realm_transform_send_error(rooted_error.handle(), can_gc);

--- a/components/script/dom/readablestreamdefaultcontroller.rs
+++ b/components/script/dom/readablestreamdefaultcontroller.rs
@@ -13,6 +13,7 @@ use js::jsval::{JSVal, UndefinedValue};
 use js::rust::wrappers::JS_GetPendingException;
 use js::rust::{HandleObject, HandleValue as SafeHandleValue, HandleValue, MutableHandleValue};
 use js::typedarray::Uint8;
+use script_bindings::conversions::SafeToJSValConvertible;
 
 use super::bindings::codegen::Bindings::QueuingStrategyBinding::QueuingStrategySize;
 use super::bindings::root::Dom;
@@ -30,7 +31,6 @@ use crate::dom::promisenativehandler::{Callback, PromiseNativeHandler};
 use crate::dom::readablestream::ReadableStream;
 use crate::dom::readablestreamdefaultreader::ReadRequest;
 use crate::dom::underlyingsourcecontainer::{UnderlyingSourceContainer, UnderlyingSourceType};
-use crate::js::conversions::ToJSValConvertible;
 use crate::realms::{InRealm, enter_realm};
 use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 
@@ -147,18 +147,15 @@ impl EnqueuedValue {
         }
     }
 
-    #[allow(unsafe_code)]
     fn to_jsval(&self, cx: SafeJSContext, rval: MutableHandleValue, can_gc: CanGc) {
         match self {
             EnqueuedValue::Native(chunk) => {
                 rooted!(in(*cx) let mut array_buffer_ptr = ptr::null_mut::<JSObject>());
                 create_buffer_source::<Uint8>(cx, chunk, array_buffer_ptr.handle_mut(), can_gc)
                     .expect("failed to create buffer source for native chunk.");
-                unsafe { array_buffer_ptr.to_jsval(*cx, rval) };
+                array_buffer_ptr.safe_to_jsval(cx, rval);
             },
-            EnqueuedValue::Js(value_with_size) => unsafe {
-                value_with_size.value.to_jsval(*cx, rval);
-            },
+            EnqueuedValue::Js(value_with_size) => value_with_size.value.safe_to_jsval(cx, rval),
             EnqueuedValue::CloseSentinel => {
                 unreachable!("The close sentinel is never made available as a js val.")
             },

--- a/components/script/dom/rtcdatachannel.rs
+++ b/components/script/dom/rtcdatachannel.rs
@@ -195,27 +195,27 @@ impl RTCDataChannel {
 
     #[allow(unsafe_code)]
     pub(crate) fn on_message(&self, channel_message: DataChannelMessage, can_gc: CanGc) {
-        unsafe {
-            let global = self.global();
-            let cx = GlobalScope::get_cx();
-            let _ac = JSAutoRealm::new(*cx, self.reflector().get_jsobject().get());
-            rooted!(in(*cx) let mut message = UndefinedValue());
+        let global = self.global();
+        let cx = GlobalScope::get_cx();
+        let _ac = JSAutoRealm::new(*cx, self.reflector().get_jsobject().get());
+        rooted!(in(*cx) let mut message = UndefinedValue());
 
-            match channel_message {
-                DataChannelMessage::Text(text) => {
-                    text.safe_to_jsval(cx, message.handle_mut());
+        match channel_message {
+            DataChannelMessage::Text(text) => {
+                text.safe_to_jsval(cx, message.handle_mut());
+            },
+            DataChannelMessage::Binary(data) => match &**self.binary_type.borrow() {
+                "blob" => {
+                    let blob = Blob::new(
+                        &global,
+                        BlobImpl::new_from_bytes(data, "".to_owned()),
+                        can_gc,
+                    );
+                    blob.safe_to_jsval(cx, message.handle_mut());
                 },
-                DataChannelMessage::Binary(data) => match &**self.binary_type.borrow() {
-                    "blob" => {
-                        let blob = Blob::new(
-                            &global,
-                            BlobImpl::new_from_bytes(data, "".to_owned()),
-                            can_gc,
-                        );
-                        blob.safe_to_jsval(cx, message.handle_mut());
-                    },
-                    "arraybuffer" => {
-                        rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
+                "arraybuffer" => {
+                    rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
+                    unsafe {
                         assert!(
                             ArrayBuffer::create(
                                 *cx,
@@ -223,24 +223,24 @@ impl RTCDataChannel {
                                 array_buffer.handle_mut()
                             )
                             .is_ok()
-                        );
+                        )
+                    };
 
-                        (*array_buffer).safe_to_jsval(cx, message.handle_mut());
-                    },
-                    _ => unreachable!(),
+                    (*array_buffer).safe_to_jsval(cx, message.handle_mut());
                 },
-            }
-
-            MessageEvent::dispatch_jsval(
-                self.upcast(),
-                &global,
-                message.handle(),
-                Some(&global.origin().immutable().ascii_serialization()),
-                None,
-                vec![],
-                can_gc,
-            );
+                _ => unreachable!(),
+            },
         }
+
+        MessageEvent::dispatch_jsval(
+            self.upcast(),
+            &global,
+            message.handle(),
+            Some(&global.origin().immutable().ascii_serialization()),
+            None,
+            vec![],
+            can_gc,
+        );
     }
 
     pub(crate) fn on_state_change(&self, state: DataChannelState, can_gc: CanGc) {

--- a/components/script/dom/rtcdatachannel.rs
+++ b/components/script/dom/rtcdatachannel.rs
@@ -7,11 +7,11 @@ use std::ptr;
 
 use constellation_traits::BlobImpl;
 use dom_struct::dom_struct;
-use js::conversions::ToJSValConvertible;
 use js::jsapi::{JSAutoRealm, JSObject};
 use js::jsval::UndefinedValue;
 use js::rust::CustomAutoRooterGuard;
 use js::typedarray::{ArrayBuffer, ArrayBufferView, CreateWith};
+use script_bindings::conversions::SafeToJSValConvertible;
 use script_bindings::weakref::WeakRef;
 use servo_media::webrtc::{
     DataChannelId, DataChannelInit, DataChannelMessage, DataChannelState, WebRtcError,
@@ -203,7 +203,7 @@ impl RTCDataChannel {
 
             match channel_message {
                 DataChannelMessage::Text(text) => {
-                    text.to_jsval(*cx, message.handle_mut());
+                    text.safe_to_jsval(cx, message.handle_mut());
                 },
                 DataChannelMessage::Binary(data) => match &**self.binary_type.borrow() {
                     "blob" => {
@@ -212,7 +212,7 @@ impl RTCDataChannel {
                             BlobImpl::new_from_bytes(data, "".to_owned()),
                             can_gc,
                         );
-                        blob.to_jsval(*cx, message.handle_mut());
+                        blob.safe_to_jsval(cx, message.handle_mut());
                     },
                     "arraybuffer" => {
                         rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
@@ -225,7 +225,7 @@ impl RTCDataChannel {
                             .is_ok()
                         );
 
-                        (*array_buffer).to_jsval(*cx, message.handle_mut());
+                        (*array_buffer).safe_to_jsval(cx, message.handle_mut());
                     },
                     _ => unreachable!(),
                 },

--- a/components/script/dom/websocket.rs
+++ b/components/script/dom/websocket.rs
@@ -590,24 +590,24 @@ impl TaskOnce for MessageReceivedTask {
 
         // Step 2-5.
         let global = ws.global();
-        // GlobalScope::get_cx() returns a valid `JSContext` pointer, so this is safe.
-        unsafe {
-            let cx = GlobalScope::get_cx();
-            let _ac = JSAutoRealm::new(*cx, ws.reflector().get_jsobject().get());
-            rooted!(in(*cx) let mut message = UndefinedValue());
-            match self.message {
-                MessageData::Text(text) => text.safe_to_jsval(cx, message.handle_mut()),
-                MessageData::Binary(data) => match ws.binary_type.get() {
-                    BinaryType::Blob => {
-                        let blob = Blob::new(
-                            &global,
-                            BlobImpl::new_from_bytes(data, "".to_owned()),
-                            CanGc::note(),
-                        );
-                        blob.safe_to_jsval(cx, message.handle_mut());
-                    },
-                    BinaryType::Arraybuffer => {
-                        rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
+        let cx = GlobalScope::get_cx();
+        let _ac = JSAutoRealm::new(*cx, ws.reflector().get_jsobject().get());
+        rooted!(in(*cx) let mut message = UndefinedValue());
+        match self.message {
+            MessageData::Text(text) => text.safe_to_jsval(cx, message.handle_mut()),
+            MessageData::Binary(data) => match ws.binary_type.get() {
+                BinaryType::Blob => {
+                    let blob = Blob::new(
+                        &global,
+                        BlobImpl::new_from_bytes(data, "".to_owned()),
+                        CanGc::note(),
+                    );
+                    blob.safe_to_jsval(cx, message.handle_mut());
+                },
+                BinaryType::Arraybuffer => {
+                    rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
+                    // GlobalScope::get_cx() returns a valid `JSContext` pointer, so this is safe.
+                    unsafe {
                         assert!(
                             ArrayBuffer::create(
                                 *cx,
@@ -615,21 +615,21 @@ impl TaskOnce for MessageReceivedTask {
                                 array_buffer.handle_mut()
                             )
                             .is_ok()
-                        );
+                        )
+                    };
 
-                        (*array_buffer).safe_to_jsval(cx, message.handle_mut());
-                    },
+                    (*array_buffer).safe_to_jsval(cx, message.handle_mut());
                 },
-            }
-            MessageEvent::dispatch_jsval(
-                ws.upcast(),
-                &global,
-                message.handle(),
-                Some(&ws.origin().ascii_serialization()),
-                None,
-                vec![],
-                CanGc::note(),
-            );
+            },
         }
+        MessageEvent::dispatch_jsval(
+            ws.upcast(),
+            &global,
+            message.handle(),
+            Some(&ws.origin().ascii_serialization()),
+            None,
+            vec![],
+            CanGc::note(),
+        );
     }
 }

--- a/components/script/dom/websocket.rs
+++ b/components/script/dom/websocket.rs
@@ -22,13 +22,13 @@ use net_traits::{
     CoreResourceMsg, FetchChannels, MessageData, WebSocketDomAction, WebSocketNetworkEvent,
 };
 use profile_traits::ipc as ProfiledIpc;
+use script_bindings::conversions::SafeToJSValConvertible;
 use servo_url::{ImmutableOrigin, ServoUrl};
 
 use crate::dom::bindings::cell::DomRefCell;
 use crate::dom::bindings::codegen::Bindings::BlobBinding::BlobMethods;
 use crate::dom::bindings::codegen::Bindings::WebSocketBinding::{BinaryType, WebSocketMethods};
 use crate::dom::bindings::codegen::UnionTypes::StringOrStringSequence;
-use crate::dom::bindings::conversions::ToJSValConvertible;
 use crate::dom::bindings::error::{Error, ErrorResult, Fallible};
 use crate::dom::bindings::inheritance::Castable;
 use crate::dom::bindings::refcounted::Trusted;
@@ -596,7 +596,7 @@ impl TaskOnce for MessageReceivedTask {
             let _ac = JSAutoRealm::new(*cx, ws.reflector().get_jsobject().get());
             rooted!(in(*cx) let mut message = UndefinedValue());
             match self.message {
-                MessageData::Text(text) => text.to_jsval(*cx, message.handle_mut()),
+                MessageData::Text(text) => text.safe_to_jsval(cx, message.handle_mut()),
                 MessageData::Binary(data) => match ws.binary_type.get() {
                     BinaryType::Blob => {
                         let blob = Blob::new(
@@ -604,7 +604,7 @@ impl TaskOnce for MessageReceivedTask {
                             BlobImpl::new_from_bytes(data, "".to_owned()),
                             CanGc::note(),
                         );
-                        blob.to_jsval(*cx, message.handle_mut());
+                        blob.safe_to_jsval(cx, message.handle_mut());
                     },
                     BinaryType::Arraybuffer => {
                         rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
@@ -617,7 +617,7 @@ impl TaskOnce for MessageReceivedTask {
                             .is_ok()
                         );
 
-                        (*array_buffer).to_jsval(*cx, message.handle_mut());
+                        (*array_buffer).safe_to_jsval(cx, message.handle_mut());
                     },
                 },
             }

--- a/components/script/dom/webxr/xrinputsource.rs
+++ b/components/script/dom/webxr/xrinputsource.rs
@@ -4,10 +4,10 @@
 
 use dom_struct::dom_struct;
 use embedder_traits::GamepadSupportedHapticEffects;
-use js::conversions::ToJSValConvertible;
 use js::jsapi::Heap;
 use js::jsval::{JSVal, UndefinedValue};
 use js::rust::MutableHandleValue;
+use script_bindings::conversions::SafeToJSValConvertible;
 use webxr_api::{Handedness, InputFrame, InputId, InputSource, TargetRayMode};
 
 use crate::dom::bindings::codegen::Bindings::XRInputSourceBinding::{
@@ -73,7 +73,6 @@ impl XRInputSource {
         }
     }
 
-    #[allow(unsafe_code)]
     pub(crate) fn new(
         window: &Window,
         session: &XRSession,
@@ -88,11 +87,12 @@ impl XRInputSource {
 
         let _ac = enter_realm(window);
         let cx = GlobalScope::get_cx();
-        unsafe {
-            rooted!(in(*cx) let mut profiles = UndefinedValue());
-            source.info.profiles.to_jsval(*cx, profiles.handle_mut());
-            source.profiles.set(profiles.get());
-        }
+        rooted!(in(*cx) let mut profiles = UndefinedValue());
+        source
+            .info
+            .profiles
+            .safe_to_jsval(cx, profiles.handle_mut());
+        source.profiles.set(profiles.get());
         source
     }
 

--- a/components/script/dom/webxr/xrviewerpose.rs
+++ b/components/script/dom/webxr/xrviewerpose.rs
@@ -4,10 +4,10 @@
 
 use dom_struct::dom_struct;
 use euclid::RigidTransform3D;
-use js::conversions::ToJSValConvertible;
 use js::jsapi::Heap;
 use js::jsval::{JSVal, UndefinedValue};
 use js::rust::MutableHandleValue;
+use script_bindings::conversions::SafeToJSValConvertible;
 use webxr_api::{Viewer, ViewerPose, Views};
 
 use crate::dom::bindings::codegen::Bindings::XRViewBinding::XREye;
@@ -38,7 +38,6 @@ impl XRViewerPose {
         }
     }
 
-    #[allow(unsafe_code)]
     pub(crate) fn new(
         window: &Window,
         session: &XRSession,
@@ -183,11 +182,9 @@ impl XRViewerPose {
         );
 
         let cx = GlobalScope::get_cx();
-        unsafe {
-            rooted!(in(*cx) let mut jsval = UndefinedValue());
-            views.to_jsval(*cx, jsval.handle_mut());
-            pose.views.set(jsval.get());
-        }
+        rooted!(in(*cx) let mut jsval = UndefinedValue());
+        views.safe_to_jsval(cx, jsval.handle_mut());
+        pose.views.set(jsval.get());
 
         pose
     }

--- a/components/script/dom/window.rs
+++ b/components/script/dom/window.rs
@@ -40,7 +40,6 @@ use euclid::default::{Point2D as UntypedPoint2D, Rect as UntypedRect, Size2D as 
 use euclid::{Point2D, Scale, Size2D, Vector2D};
 use fonts::FontContext;
 use ipc_channel::ipc::{self, IpcSender};
-use js::conversions::ToJSValConvertible;
 use js::glue::DumpJSStack;
 use js::jsapi::{
     GCReason, Heap, JS_GC, JSAutoRealm, JSContext as RawJSContext, JSObject, JSPROP_ENUMERATE,
@@ -70,6 +69,7 @@ use profile_traits::mem::ProfilerChan as MemProfilerChan;
 use profile_traits::time::ProfilerChan as TimeProfilerChan;
 use script_bindings::codegen::GenericBindings::NavigatorBinding::NavigatorMethods;
 use script_bindings::codegen::GenericBindings::PerformanceBinding::PerformanceMethods;
+use script_bindings::conversions::SafeToJSValConvertible;
 use script_bindings::interfaces::WindowHelpers;
 use script_bindings::root::Root;
 use script_traits::ScriptThreadMessage;
@@ -1789,12 +1789,9 @@ impl WindowMethods<crate::DomTypeHolder> for Window {
     }
 
     // https://dom.spec.whatwg.org/#dom-window-event
-    #[allow(unsafe_code)]
     fn Event(&self, cx: JSContext, rval: MutableHandleValue) {
         if let Some(ref event) = *self.current_event.borrow() {
-            unsafe {
-                event.reflector().get_jsobject().to_jsval(*cx, rval);
-            }
+            event.reflector().get_jsobject().safe_to_jsval(cx, rval);
         }
     }
 

--- a/components/script/dom/writablestream.rs
+++ b/components/script/dom/writablestream.rs
@@ -18,6 +18,7 @@ use js::rust::{
     MutableHandleValue as SafeMutableHandleValue,
 };
 use script_bindings::codegen::GenericBindings::MessagePortBinding::MessagePortMethods;
+use script_bindings::conversions::SafeToJSValConvertible;
 
 use super::bindings::codegen::Bindings::QueuingStrategyBinding::QueuingStrategySize;
 use crate::dom::bindings::cell::DomRefCell;
@@ -41,7 +42,6 @@ use crate::dom::writablestreamdefaultcontroller::{
     UnderlyingSinkType, WritableStreamDefaultController,
 };
 use crate::dom::writablestreamdefaultwriter::WritableStreamDefaultWriter;
-use crate::js::conversions::ToJSValConvertible;
 use crate::realms::{InRealm, enter_realm};
 use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 
@@ -1191,7 +1191,6 @@ impl CrossRealmTransformWritable {
 
     /// <https://streams.spec.whatwg.org/#abstract-opdef-setupcrossrealmtransformwritable>
     /// Add a handler for port’s messageerror event with the following steps:
-    #[allow(unsafe_code)]
     pub(crate) fn handle_error(
         &self,
         cx: SafeJSContext,
@@ -1203,7 +1202,7 @@ impl CrossRealmTransformWritable {
         // Let error be a new "DataCloneError" DOMException.
         let error = DOMException::new(global, DOMErrorName::DataCloneError, can_gc);
         rooted!(in(*cx) let mut rooted_error = UndefinedValue());
-        unsafe { error.to_jsval(*cx, rooted_error.handle_mut()) };
+        error.safe_to_jsval(cx, rooted_error.handle_mut());
 
         // Perform ! CrossRealmTransformSendError(port, error).
         port.cross_realm_transform_send_error(rooted_error.handle(), can_gc);

